src/main/scala/edu/alexy/lectures/concurrency/RoboAdviser.scala

```diff
@@ -4,48 +4,113 @@ import edu.alexy.lectures.concurrency.model.Company
 import edu.alexy.lectures.concurrency.util.{Db, DbError, WebApi, WebApiError}
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
+import scala.util.{Failure, Success}
 
 object RoboAdviser {
   // Task 1.
   // Return 'AAPL' revenue from `Db.getCompanyLastFinancials`. Possible error should be returned as a ServiceError.
-  def getAAPLRevenue: Future[Double] = ???
+  def getAAPLRevenue: Future[Double] =
+    Db.getCompanyLastFinancials("AAPL")
+      .transformWith {
+        case Success(Some(financials)) => Future.successful(financials.revenue)
+        case _                         => Future.failed(DbError)
+      }
+
+  private def futureRetry[T](block: => Future[T], retries: Int): Future[T] = {
+    val result: Future[T] = block
+
+    result.recoverWith {
+      case _ if retries > 1 => futureRetry[T](block, retries - 1)
+    }
+  }
 
   // Task 2.
   // Implement a fallback strategy for 'Db.getAllTickers'.
   // 'Db.getAllTickers' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getAllTickersRetryable(retries: Int = 10): Future[Seq[String]] = ???
+  def getAllTickersRetryable(retries: Int = 10): Future[Seq[String]] =
+    futureRetry(Db.getAllTickers, retries).transform {
+      case Failure(exception) => Failure(DbError)
+      case other              => other
+    }
 
   // Task 3.
   // Implement a fallback strategy for 'Db.getCompanyLastFinancials'.
   // 'Db.getCompanyLastFinancials' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getCompanyRetryable(ticker: String, retries: Int = 10): Future[Option[Company]] = ???
+  def getCompanyRetryable(ticker: String,
+                          retries: Int = 10): Future[Option[Company]] =
+    futureRetry(Db.getCompanyLastFinancials(ticker), retries).transform {
+      case Failure(exception) => Failure(DbError)
+      case other              => other
+    }
 
   // Task 4.
   // Implement a fallback strategy 'WebApi.getPrice'.
   // 'WebApi.getPrice' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] = ???
+  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] =
+    futureRetry(WebApi.getPrice(ticker), retries).transform {
+      case Failure(exception) => Failure(WebApiError)
+      case other              => other
+    }
 
   // Task 5.
   // Using retryable functions return all tickers with their real time prices.
-  def getAllTickersPrices: Future[Seq[(String, Double)]] = ???
+  def getAllTickersPrices: Future[Seq[(String, Double)]] =
+    for (tickers <- getAllTickersRetryable();
+         tikerWithPrice <- Future.traverse(tickers) { ticker =>
+           Future.successful(ticker).zip(getPriceRetryable(ticker))
+         }) yield tikerWithPrice
```

Comment:
If you need to write more than one line, just use { } :
 

```
for {
      tickers <- getAllTickersRetryable()
      tickerWithPrice <- Future.traverse(tickers)(ticker => getPriceRetryable(ticker).map(ticker -> _))
    }  yield tickerWithPrice

```

As a rule of thumb, you should always use {} in for-exp. If for-exp contains less than two lines then you should use `map`(or `foreach`) instead.